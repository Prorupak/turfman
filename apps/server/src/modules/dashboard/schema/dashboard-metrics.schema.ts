import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { HydratedDocument } from 'mongoose';

export type DashboardMetricsDocument = HydratedDocument<DashboardMetrics>;

@Schema({
  collection: 'dashboard_metrics',
  timestamps: { createdAt: 'createdAt', updatedAt: 'updatedAt' },
})
export class DashboardMetrics {
  @ApiProperty({
    description: 'The total number of returned products.',
    example: 50,
  })
  @Prop({ default: 0 })
  totalReturns: number;

  @ApiProperty({
    description: 'The rate of product returns as a percentage.',
    example: 5.0,
  })
  @Prop({ default: 0 })
  returnRate: number;

  @ApiProperty({
    description: 'The total revenue impact of returned products.',
    example: 5000.0,
  })
  @Prop({ default: 0 })
  totalRevenueImpact: number;

  @ApiPropertyOptional({
    description:
      'An object representing reasons for product returns and their counts.',
    example: { 'Damaged product': 10, 'Wrong item': 5 },
    type: Object,
  })
  @Prop({ type: Map, of: Number, default: {} })
  returnReasons: Map<string, number>;

  @ApiPropertyOptional({
    description: 'The date when metrics were last updated.',
    example: '2024-08-29T03:43:32.702Z',
  })
  @Prop({ type: Date, default: Date.now })
  updatedAt: Date;
}

export const DashboardMetricsSchema =
  SchemaFactory.createForClass(DashboardMetrics);
