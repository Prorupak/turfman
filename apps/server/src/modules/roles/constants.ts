export const roleSelect = {
  _id: 0,
  id: '$_id',
  name: true,
  sort: true,
  default: true,
  createdAt: true,
  updatedAt: true,
};

export enum UserRoles {
  SALES_ASSISTANCE = 'sales_assistance',
  ADMIN = 'admin',
  SUPER_ADMIN = 'super_admin',
}
