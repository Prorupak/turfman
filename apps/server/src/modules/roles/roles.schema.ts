// src/schemas/role.schema.ts

import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Document, Types } from 'mongoose';

export type RoleDocument = Role & Document;

@Schema({
  collection: 'roles',
  timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' },
})
export class Role extends Document {
  @Prop({
    type: String,
    required: true,
    unique: true,
    default: () => new Types.ObjectId(),
  })
  id: string;

  @Prop({
    type: String,
    required: [true, 'Role name is required'],
    unique: true,
    minlength: [2, 'Role name must be at least 2 characters long'],
    maxlength: [255, 'Role name cannot exceed 255 characters'],
  })
  name: string;

  @Prop({ type: Number, required: true })
  sort: number;

  @Prop({ type: Boolean, default: false })
  default: boolean;

  @Prop({ type: Types.ObjectId, ref: 'UserRole' })
  userRoles: Types.ObjectId[];
}

export const RoleSchema = SchemaFactory.createForClass(Role);

// Adding an index for sort
RoleSchema.index({ sort: 1 });
