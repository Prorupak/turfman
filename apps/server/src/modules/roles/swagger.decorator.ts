import { applyDecorators } from '@nestjs/common';
import {
  ApiBadRequestResponse,
  ApiBody,
  ApiNotFoundResponse,
  ApiOkResponse,
  ApiOperation,
  ApiParam,
  ApiResponse,
} from '@nestjs/swagger';
import { RoleDto } from './dtos/role.dto';

export const GetAllRolesDocs = () => {
  return applyDecorators(
    ApiOperation({ summary: 'Get all roles' }),
    ApiResponse({
      status: 200,
      description: 'List of all roles',
      type: [RoleDto],
    }),
  );
};

export const GetRoleByIdDocs = () => {
  return applyDecorators(
    ApiOperation({ summary: 'Get role by ID' }),
    ApiParam({
      name: 'id',
      description:
        'The unique identifier of the role, should be a valid MongoDB ObjectId',
      example: '60b8d2958e4e1f001f8f2a5d',
    }),
    ApiResponse({
      status: 200,
      description: 'Details of the role',
      type: RoleDto,
    }),
    ApiNotFoundResponse({
      status: 404,
      schema: {
        type: 'object',
        properties: {
          statusCode: { type: 'number', example: 404 },
          message: { type: 'string', example: 'Entity not found' },
        },
      },
    }),
  );
};

export const CreateRoleDocs = () => {
  return applyDecorators(
    ApiOperation({ summary: 'Create role' }),
    ApiBody({
      description: 'Request body data',
      schema: {
        type: 'object',
        properties: {
          name: { type: 'string', example: 'Admin' },
        },
      },
    }),
    ApiResponse({
      status: 200,
      description: 'Created details of the role',
      type: RoleDto,
    }),
  );
};

export const UpdateRoleDocs = () => {
  return applyDecorators(
    ApiOperation({ summary: 'Create role' }),
    ApiParam({
      name: 'id',
      description:
        'The unique identifier of the role, should be a valid MongoDB ObjectId',
      example: '60b8d2958e4e1f001f8f2a5d',
    }),
    ApiBody({
      description: 'Request body data',
      schema: {
        type: 'object',
        properties: {
          name: { type: 'string', example: 'Admin' },
          id: { type: 'string', example: '60b8d2958e4e1f001f8f2a5d' },
        },
      },
    }),
    ApiResponse({
      status: 200,
      description: 'Updated details of the role',
      type: RoleDto,
    }),
  );
};

export const DeleteRoleDocs = () => {
  return applyDecorators(
    ApiOperation({ summary: 'Delete a role by ID' }),
    ApiParam({
      name: 'id',
      description:
        'The unique identifier of the role to delete, should be a valid MongoDB ObjectId',
      example: '60b8d2958e4e1f001f8f2a5d',
    }),
    ApiBody({
      description: 'Request body data',
      schema: {
        type: 'object',
        properties: {
          id: { type: 'string', example: '60b8d2958e4e1f001f8f2a5d' },
        },
      },
    }),
    ApiResponse({
      status: 200,
      description: 'The role has been successfully deleted.',
      type: RoleDto,
    }),
    ApiNotFoundResponse({
      status: 404,
      schema: {
        type: 'object',
        properties: {
          statusCode: { type: 'number', example: 404 },
          message: { type: 'string', example: 'Entity not found' },
        },
      },
    }),
    ApiBadRequestResponse({
      status: 400,
      schema: {
        type: 'object',
        properties: {
          statusCode: { type: 'number', example: 400 },
          message: { type: 'string', example: 'Cannot delete a default role' },
        },
      },
    }),
  );
};

export const SortRoleDocs = () => {
  return applyDecorators(
    ApiOperation({ summary: 'Update the sort order of a role by ID' }),
    ApiParam({
      name: 'id',
      description:
        'The unique identifier of the role to update, should be a valid MongoDB ObjectId',
      example: '60b8d2958e4e1f001f8f2a5d',
    }),
    ApiBody({
      description: 'Request body data',
      schema: {
        type: 'object',
        properties: {
          id: { type: 'string', example: '60b8d2958e4e1f001f8f2a5d' },
          sort: { type: 'number', example: 2 },
        },
      },
    }),
    ApiOkResponse({
      description: 'The role with updated sort order',
      type: RoleDto,
    }),
    ApiResponse({
      status: 400,
      description: 'Invalid sort order or no change detected',
    }),
    ApiBadRequestResponse({
      schema: {
        type: 'object',
        properties: {
          statusCode: { type: 'number', example: 400 },
          message: {
            type: 'string',
            example: 'Invalid sort order or no change detected',
          },
        },
      },
    }),
    ApiNotFoundResponse({
      status: 404,
      schema: {
        type: 'object',
        properties: {
          statusCode: { type: 'number', example: 404 },
          message: { type: 'string', example: 'Entity not found' },
        },
      },
    }),
  );
};
