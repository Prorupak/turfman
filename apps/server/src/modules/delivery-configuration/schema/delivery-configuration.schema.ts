import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { HydratedDocument, Types } from 'mongoose';
import { Category } from 'modules/category/schemas/category.schema'; // Import the Category schema

export type DeliveryConfigurationDocument =
  HydratedDocument<DeliveryConfiguration>;

@Schema({
  collection: 'delivery_configurations',
  timestamps: { createdAt: 'createdAt', updatedAt: 'updatedAt' },
})
export class DeliveryConfiguration {
  @ApiProperty({
    description:
      'Reference to the category this delivery configuration applies to.',
    type: String,
    example: '66c73abbd4c3c13c85b395e4',
  })
  @Prop({ type: Types.ObjectId, ref: Category.name, required: true })
  category: Types.ObjectId; // Reference to the Category model

  @ApiPropertyOptional({
    description: 'Flat rate for delivery if applicable.',
    example: 50.0,
  })
  @Prop({ required: false })
  flatRate?: number;

  @ApiPropertyOptional({
    description:
      'Region-specific delivery rates, with regions as keys and rates as values.',
    example: { NY: 75.0, CA: 100.0 },
    type: 'object',
    additionalProperties: { type: 'number' },
  })
  @Prop({ type: Map, of: Number })
  regionSpecificRates?: Record<string, number>;

  @ApiPropertyOptional({
    description:
      'List of applicable postcodes where this delivery configuration applies.',
    example: ['10001', '20002', '94101'],
    type: [String],
  })
  @Prop([String || Number])
  applicablePostcodes?: Array<string | number>;

  @ApiProperty({
    description: 'Indicates if the delivery configuration is active.',
    example: true,
    default: true,
  })
  @Prop({ default: true })
  isActive: boolean;

  @ApiPropertyOptional({
    description: 'Creation date of config.',
    example: 'Thu Aug 22 2024 21:29:50 GMT+0545 (Nepal Time)',
    type: Date,
  })
  @Prop({ type: Date, default: Date.now() })
  createdAt?: Date;

  @ApiPropertyOptional({
    description: 'Updated date of config.',
    example: 'Thu Aug 22 2024 21:29:50 GMT+0545 (Nepal Time)',
    type: Date,
  })
  @Prop({ type: Date, default: Date.now() })
  updatedAt?: Date;
}

export const DeliveryConfigurationSchema = SchemaFactory.createForClass(
  DeliveryConfiguration,
);

DeliveryConfigurationSchema.index({ createdAt: -1 });
DeliveryConfigurationSchema.index({ updatedAt: -1 });
DeliveryConfigurationSchema.index({ category: 1 });
DeliveryConfigurationSchema.index({ flatRate: 1 });
