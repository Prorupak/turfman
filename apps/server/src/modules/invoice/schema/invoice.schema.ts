import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { Types, HydratedDocument } from 'mongoose';
import { ApiProperty } from '@nestjs/swagger';
import { Order } from 'modules/order-management/schemas/order-management.schema';

export type InvoiceDocument = HydratedDocument<Invoice>;

export enum InvoiceStatus {
  UNPAID = 'UNPAID',
  PARTIALLY_PAID = 'PARTIALLY_PAID',
  PENDING = 'PENDING',
  PAID = 'PAID',
}

@Schema({
  collection: 'invoices',
  timestamps: { createdAt: 'createdAt', updatedAt: 'updatedAt' },
})
export class Invoice {
  @ApiProperty({
    description: 'Associated order ID.',
    example: '66c777594aa571f347476767',
  })
  @Prop({ type: Types.ObjectId, ref: Order.name, required: true })
  orderId: Types.ObjectId;

  @ApiProperty({
    description: 'Invoice number.',
    example: 'INV-2024-0001',
  })
  @Prop({ required: true, unique: true })
  invoiceNumber: string;

  @ApiProperty({
    description: 'Customer email.',
    example: 'customer@example.com',
  })
  @Prop({ required: true })
  customerEmail: string;

  @ApiProperty({
    description: 'Total amount of the invoice.',
    example: 150.75,
  })
  @Prop({ required: true })
  totalAmount: number;

  @ApiProperty({
    description: 'Payment status of the invoice.',
    enum: InvoiceStatus,
    example: InvoiceStatus.UNPAID,
  })
  @Prop({ required: true, enum: InvoiceStatus, default: InvoiceStatus.PENDING })
  status: InvoiceStatus;

  @ApiProperty({
    description: 'Due date for payment.',
    example: '2024-09-30T00:00:00Z',
  })
  @Prop({ type: Date, required: true })
  dueDate: Date;

  @ApiProperty({
    description: 'Payment details (method, transaction ID, etc.).',
    example: { method: 'Credit Card', transactionId: 'XYZ123' },
    type: 'object',
  })
  @Prop({ type: Map, of: String })
  paymentDetails: Map<string, string>;

  @ApiProperty({
    description: 'Creation date of the invoice.',
    example: 'Thu Aug 22 2024 21:29:50 GMT+0545 (Nepal Time)',
    type: Date,
  })
  @Prop({ type: Date, default: Date.now })
  createdAt?: Date;

  @ApiProperty({
    description: 'Updated date of the invoice.',
    example: 'Thu Aug 22 2024 21:29:50 GMT+0545 (Nepal Time)',
    type: Date,
  })
  @Prop({ type: Date, default: Date.now })
  updatedAt?: Date;
}

export const InvoiceSchema = SchemaFactory.createForClass(Invoice);

InvoiceSchema.index({ createdAt: -1 });
InvoiceSchema.index({ updatedAt: -1 });
InvoiceSchema.index({ orderId: 1 });
InvoiceSchema.index({ customerEmail: 1 });
