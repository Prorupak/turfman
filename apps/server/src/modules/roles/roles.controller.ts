import {
  Body,
  Controller,
  Delete,
  Get,
  Patch,
  Post,
  Put,
  Query,
  UseInterceptors,
} from '@nestjs/common';
import {
  CreateRoleDto,
  DeleteRoleDto,
  SearchRoleDto,
  SortRoleDto,
  UpdateRoleDto,
} from './dtos';
import { RolesService } from './roles.service';
import { Roles } from 'decorators/auth';
import { RolesInterceptor } from 'interceptors/auth';
import { ApiName } from 'decorators/openapi';
import {
  CreateRoleDocs,
  DeleteRoleDocs,
  GetAllRolesDocs,
  GetRoleByIdDocs,
  SortRoleDocs,
  UpdateRoleDocs,
} from './swagger.decorator';
import { SecureEndpoint } from 'guards';
import { UserRoles } from './constants';

@Roles(UserRoles.ADMIN, UserRoles.SUPER_ADMIN)
@ApiName()
@Controller('roles')
@UseInterceptors(RolesInterceptor)
export class RolesController {
  constructor(private rolesService: RolesService) {}

  @Get()
  @SecureEndpoint.apply()
  @GetAllRolesDocs()
  async search() {
    return this.rolesService.findAll();
  }

  @Get(':id')
  @SecureEndpoint.apply()
  @GetRoleByIdDocs()
  async getById(@Query() dto: SearchRoleDto) {
    return this.rolesService.findById(dto);
  }

  @Post()
  @SecureEndpoint.apply()
  @CreateRoleDocs()
  async create(@Body() dto: CreateRoleDto) {
    return this.rolesService.create(dto);
  }

  @Put(':id')
  @SecureEndpoint.apply()
  @UpdateRoleDocs()
  async update(@Body() dto: UpdateRoleDto) {
    return this.rolesService.update(dto);
  }

  @Delete(':id')
  @SecureEndpoint.apply()
  @DeleteRoleDocs()
  async delete(@Body() dto: DeleteRoleDto) {
    return this.rolesService.delete(dto);
  }

  @Patch(':id/sort')
  @SecureEndpoint.apply()
  @SortRoleDocs()
  async sort(@Body() dto: SortRoleDto) {
    return this.rolesService.sort(dto);
  }
}
