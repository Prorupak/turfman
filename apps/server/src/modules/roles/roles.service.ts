import { Injectable, Logger } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { AppError } from 'src/common/errors';
import {
  CreateRoleDto,
  DeleteRoleDto,
  SearchRoleDto,
  SortRoleDto,
  UpdateRoleDto,
} from './dtos';
import { Role, RoleDocument } from './roles.schema';
import { roleSelect } from './constants';
import { messages } from 'src/constants/messages';

@Injectable()
export class RolesService {
  private logger = new Logger(RolesService.name);

  constructor(@InjectModel(Role.name) private roleModel: Model<RoleDocument>) {}

  async findAll() {
    return this.roleModel
      .find()
      .sort({ sort: 'asc' })
      .select(roleSelect)
      .exec();
  }

  async findById(dto: SearchRoleDto) {
    const role = await this.roleModel
      .findById(dto.id)
      .select(roleSelect)
      .exec();

    if (!role) {
      throw new AppError.NotFound(messages.error.notFoundEntity).setParams({
        entity: 'Role',
        id: dto.id,
      });
    }

    return role;
  }

  private async maxSort() {
    const result = await this.roleModel
      .findOne()
      .sort({ sort: -1 })
      .select('sort')
      .exec();
    return result ? result.sort : 0;
  }

  async create(dto: CreateRoleDto) {
    const session = await this.roleModel.startSession();
    session.startTransaction();

    try {
      const maxSort = await this.maxSort();

      const role = new this.roleModel({
        name: dto.name,
        sort: maxSort !== null ? maxSort + 1 : 1,
      });

      await role.save({ session });
      this.logger.log(`[Create]: ${role.name}`);
      await session.commitTransaction();
      return role;
    } catch (error) {
      await session.abortTransaction();
      throw error;
    } finally {
      session.endSession();
    }
  }

  async update(dto: UpdateRoleDto) {
    const session = await this.roleModel.startSession();
    session.startTransaction();

    try {
      const currentRole = await this.findById({ id: dto.id });

      if (currentRole.default) {
        throw new AppError.Argument(messages.warning.defaultData);
      }

      const role = await this.roleModel
        .findByIdAndUpdate(dto.id, { name: dto.name }, { new: true, session })
        .select(roleSelect)
        .exec();

      this.logger.log(`[Update]: ${role.name}`);
      await session.commitTransaction();
      return role;
    } catch (error) {
      await session.abortTransaction();
      throw error;
    } finally {
      session.endSession();
    }
  }

  async delete(dto: DeleteRoleDto) {
    const session = await this.roleModel.startSession();
    session.startTransaction();

    try {
      const currentRole = await this.findById({ id: dto.id });

      if (currentRole.default) {
        throw new AppError.Argument(messages.warning.defaultData);
      }

      const role = await this.roleModel
        .findByIdAndDelete(dto.id, { session })
        .select(roleSelect)
        .exec();

      await this.roleModel.updateMany(
        { sort: { $gt: role.sort } },
        { $inc: { sort: -1 } },
        { session },
      );

      this.logger.warn(`[Delete]: ${role.name}`);
      await session.commitTransaction();
      return role;
    } catch (error) {
      await session.abortTransaction();
      throw error;
    } finally {
      session.endSession();
    }
  }

  async sort(dto: SortRoleDto) {
    const session = await this.roleModel.startSession();
    session.startTransaction();

    try {
      const maxSort = await this.maxSort();

      if (dto.sort > maxSort) {
        throw new AppError.Argument(messages.error.changeSort).setErrors([
          `Max sort is ${maxSort}, current sort is ${dto.sort}`,
        ]);
      }

      const newSort = dto.sort;
      const currentRole = await this.roleModel
        .findById(dto.id)
        .select('sort')
        .exec();

      if (!currentRole) {
        throw new AppError.NotFound(messages.error.notFoundEntity).setParams({
          entity: 'Role',
          id: dto.id,
        });
      }

      const oldSort = currentRole.sort;

      if (oldSort < newSort) {
        await this.roleModel.updateMany(
          { sort: { $lte: newSort, $gt: oldSort } },
          { $inc: { sort: -1 } },
          { session },
        );
      } else if (oldSort > newSort) {
        await this.roleModel.updateMany(
          { sort: { $gte: newSort, $lt: oldSort } },
          { $inc: { sort: 1 } },
          { session },
        );
      } else {
        throw new AppError.Argument(messages.warning.noChange);
      }

      const role = await this.roleModel
        .findByIdAndUpdate(dto.id, { sort: dto.sort }, { new: true, session })
        .select(roleSelect)
        .exec();

      this.logger.warn(`[Sort]: ${role.name}`);
      await session.commitTransaction();
      return role;
    } catch (error) {
      await session.abortTransaction();
      throw error;
    } finally {
      session.endSession();
    }
  }
}
