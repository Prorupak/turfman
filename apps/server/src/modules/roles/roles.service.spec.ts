import { Test, TestingModule } from '@nestjs/testing';
import { getModelToken } from '@nestjs/mongoose';
import { RolesService } from './roles.service';
import { Role } from './roles.schema';

describe('RolesService', () => {
  let service: RolesService;
  let roleModel: any;

  const mockSession = {
    startTransaction: jest.fn(),
    commitTransaction: jest.fn(),
    abortTransaction: jest.fn(),
    endSession: jest.fn(),
  };

  const mockRoleModel = {
    create: jest.fn(), // Mock the create method
    find: jest.fn().mockReturnThis(),
    findOne: jest.fn().mockReturnThis(),
    findById: jest.fn().mockReturnThis(),
    findByIdAndUpdate: jest.fn().mockReturnThis(),
    findByIdAndDelete: jest.fn().mockReturnThis(),
    sort: jest.fn().mockReturnThis(),
    select: jest.fn().mockReturnThis(),
    lean: jest.fn().mockReturnThis(),
    exec: jest.fn(),
    startSession: jest.fn().mockResolvedValue(mockSession), // Return the mocked session object
    updateMany: jest.fn(),
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        RolesService,
        {
          provide: getModelToken(Role.name),
          useValue: mockRoleModel,
        },
      ],
    }).compile();

    service = module.get<RolesService>(RolesService);
    roleModel = module.get(getModelToken(Role.name));
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('findAll', () => {
    it('should return a list of roles', async () => {
      const mockRoles = [
        {
          name: 'Admin',
          sort: 1,
          default: false,
          createdAt: new Date(),
          updatedAt: new Date(),
          id: '1',
        },
        {
          name: 'User',
          sort: 2,
          default: false,
          createdAt: new Date(),
          updatedAt: new Date(),
          id: '2',
        },
      ];
      roleModel.exec.mockResolvedValue(mockRoles);

      const result = await service.findAll();
      expect(result).toEqual([
        {
          name: 'Admin',
          sort: 1,
          default: false,
          createdAt: mockRoles[0].createdAt,
          updatedAt: mockRoles[0].updatedAt,
          id: '1',
        },
        {
          name: 'User',
          sort: 2,
          default: false,
          createdAt: mockRoles[1].createdAt,
          updatedAt: mockRoles[1].updatedAt,
          id: '2',
        },
      ]);
    });
  });

  describe('findById', () => {
    it('should return a role by ID', async () => {
      const mockRole = {
        name: 'Admin',
        sort: 1,
        default: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        id: '1',
      };
      roleModel.exec.mockResolvedValue(mockRole);

      const result = await service.findById({ id: '1' });
      expect(result).toEqual({
        name: 'Admin',
        sort: 1,
        default: false,
        createdAt: mockRole.createdAt,
        updatedAt: mockRole.updatedAt,
        id: '1',
      });
    });

    it('should throw a NotFound error if role is not found', async () => {
      // Mock the query chain to return null (no role found)
      roleModel.exec.mockResolvedValue(null);

      await expect(
        service.findById({ id: '1' }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('maxSort', () => {
    it('should return the max sort value', async () => {
      const mockSort = { sort: 10 };
      roleModel.exec.mockResolvedValue(mockSort);

      const result = await service.maxSort();
      expect(result).toBe(10);
    });

    it('should return 0 if no sort value is found', async () => {
      roleModel.exec.mockResolvedValue(null);
      const result = await service.maxSort();
      expect(result).toBe(0);
    });
  });

  describe('create', () => {
    it('should create a role successfully', async () => {
      const createDto = { name: 'Admin' };
      const mockRole = {
        name: 'Admin',
        sort: 1,
        default: false,
        createdAt: new Date(),
        updatedAt: new Date(),
        _id: '1',
      };

      // Mock maxSort result
      roleModel.exec.mockResolvedValueOnce({ sort: 1 });

      // Mock roleModel.create to return the mockRole object
      roleModel.create.mockResolvedValue(mockRole);

      const result = await service.create(createDto);
      expect(result).toEqual({
        name: 'Admin',
        sort: 1,
        default: false,
        createdAt: mockRole.createdAt,
        updatedAt: mockRole.updatedAt,
        id: '1',
      });

      // Check if the session methods were called correctly
      expect(roleModel.startSession).toHaveBeenCalled();
      expect(mockSession.commitTransaction).toHaveBeenCalled();
      expect(mockSession.endSession).toHaveBeenCalled();
    });

    it('should rollback transaction on error', async () => {
      const createDto = { name: 'Admin' };

      // Mock maxSort result
      roleModel.exec.mockResolvedValueOnce({ sort: 1 });

      // Mock roleModel.create to throw an error
      roleModel.create.mockRejectedValue(new Error('Create failed'));

      await expect(service.create(createDto)).rejects.toThrow('Create failed');

      // Check if the session methods were called correctly
      expect(roleModel.startSession).toHaveBeenCalled();
      expect(mockSession.abortTransaction).toHaveBeenCalled();
      expect(mockSession.endSession).toHaveBeenCalled();
    });
  });

  describe('update', () => {
    it('should update a role successfully', async () => {
      const updateDto = { id: '1', name: 'User' };
      const mockRole = {
        name: 'User',
        sort: 1,
        createdAt: new Date(),
        updatedAt: new Date(),
        id: '1',
      };
      roleModel.exec.mockResolvedValueOnce(mockRole); // Mock findById
      roleModel.exec.mockResolvedValueOnce(mockRole); // Mock findByIdAndUpdate

      const result = await service.update(updateDto);
      expect(result).toEqual({
        name: 'User',
        sort: 1,
        default: undefined,
        createdAt: mockRole.createdAt,
        updatedAt: mockRole.updatedAt,
        id: '1',
      });
    });

    it('should throw an error when trying to update a default role', async () => {
      const updateDto = { id: '1', name: 'User' };
      roleModel.exec.mockResolvedValueOnce({ default: true }); // Mock findById

      await expect(
        service.update(updateDto),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should rollback transaction on error during update', async () => {
      const updateDto = { id: '1', name: 'User' };

      // Mock findById to return a non-default role
      roleModel.exec.mockResolvedValueOnce({ default: false });

      // Mock findByIdAndUpdate to throw an error
      roleModel.exec.mockRejectedValueOnce(new Error('Update failed'));

      await expect(
        service.update(updateDto),
      ).rejects.toThrowErrorMatchingSnapshot();

      // Check if the session methods were called correctly
      expect(roleModel.startSession).toHaveBeenCalled();
      expect(mockSession.abortTransaction).toHaveBeenCalled(); // Ensure abortTransaction was called
      expect(mockSession.endSession).toHaveBeenCalled();
    });
  });

  describe('delete', () => {
    it('should delete a role successfully', async () => {
      const deleteDto = { id: '1' };
      const mockRole = {
        name: 'User',
        sort: 2,
        createdAt: new Date(),
        updatedAt: new Date(),
        id: '1',
      };
      roleModel.exec.mockResolvedValueOnce(mockRole); // Mock findById
      roleModel.exec.mockResolvedValueOnce(mockRole); // Mock findByIdAndDelete

      const result = await service.delete(deleteDto);
      expect(result).toEqual(mockRole);
      expect(roleModel.updateMany).toHaveBeenCalledWith(
        { sort: { $gt: 2 } },
        { $inc: { sort: -1 } },
        expect.any(Object),
      );
    });

    it('should throw an error when trying to delete a default role', async () => {
      const deleteDto = { id: '1' };
      roleModel.exec.mockResolvedValueOnce({ default: true }); // Mock findById

      await expect(
        service.delete(deleteDto),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should rollback transaction on error during deletion', async () => {
      const deleteDto = { id: '1' };
      roleModel.exec.mockResolvedValueOnce({ default: false }); // Mock findById
      roleModel.exec.mockRejectedValue(new Error('Delete failed')); // Mock findByIdAndDelete failure

      await expect(
        service.delete(deleteDto),
      ).rejects.toThrowErrorMatchingSnapshot();
      // Check if the session methods were called correctly
      expect(roleModel.startSession).toHaveBeenCalled();
      expect(mockSession.abortTransaction).toHaveBeenCalled(); // Ensure abortTransaction was called
      expect(mockSession.endSession).toHaveBeenCalled();
    });
  });

  describe('sort', () => {
    it('should update sort order successfully', async () => {
      const sortDto = { id: '1', sort: 2 };

      const mockRole = {
        _id: '1',
        name: 'Admin',
        sort: 1,
      };

      const updatedRole = {
        _id: '1',
        name: 'Admin',
        sort: 2,
      };

      // Mock maxSort result
      roleModel.exec.mockResolvedValueOnce(3); // Simulating a maxSort of 3

      // Mock findById to return the current role
      roleModel.exec.mockResolvedValueOnce(mockRole);

      // Mock findByIdAndUpdate to return the updated role
      roleModel.exec.mockResolvedValueOnce(updatedRole);

      const result = await service.sort(sortDto);

      expect(result).toEqual(updatedRole);
      expect(roleModel.updateMany).toHaveBeenCalledWith(
        { sort: { $lte: 2, $gt: 1 } }, // Conditions for moving the sort up
        { $inc: { sort: -1 } },
        expect.any(Object),
      );
      expect(roleModel.startSession).toHaveBeenCalled();
      expect(mockSession.commitTransaction).toHaveBeenCalled();
      expect(mockSession.endSession).toHaveBeenCalled();
    });

    it('should rollback transaction on error during sort update', async () => {
      const sortDto = { id: '1', sort: 2 };
      roleModel.exec.mockResolvedValueOnce({ sort: 1 }); // Mock findById
      roleModel.exec.mockRejectedValue(new Error('Sort update failed')); // Mock updateMany or findByIdAndUpdate failure

      await expect(
        service.sort(sortDto),
      ).rejects.toThrowErrorMatchingSnapshot();
      // Check if the session methods were called correctly
      expect(roleModel.startSession).toHaveBeenCalled();
      expect(mockSession.abortTransaction).toHaveBeenCalled(); // Ensure abortTransaction was called
      expect(mockSession.endSession).toHaveBeenCalled();
    });
  });
});
