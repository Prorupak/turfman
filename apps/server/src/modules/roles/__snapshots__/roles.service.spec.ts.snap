// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RolesService delete should rollback transaction on error during deletion 1`] = `"Delete failed"`;

exports[`RolesService delete should throw an error when trying to delete a default role 1`] = `"Using default data"`;

exports[`RolesService findById should throw a NotFound error if role is not found 1`] = `"Entity not found"`;

exports[`RolesService sort should rollback transaction on error during sort update 1`] = `"Cannot change sort order"`;

exports[`RolesService sort should throw an error if the sort value exceeds the max sort 1`] = `"Delete failed"`;

exports[`RolesService sort should throw an error when sort value exceeds max sort value 1`] = `"Cannot change sort order"`;

exports[`RolesService sort should throw an error when there is no change in sort order 1`] = `"No changes were made"`;

exports[`RolesService update should rollback transaction on error during update 1`] = `"Update failed"`;

exports[`RolesService update should throw an error when trying to update a default role 1`] = `"Using default data"`;
